modules/manifold/src/constants/types.js

```diff
@@ -0,0 +1,45 @@
+import PropTypes from 'prop-types';
+import {FILTER_TYPE, FEATURE_TYPE} from '@mlvis/mlvis-common/constants';
+
+export const FIELD = PropTypes.shape({
+  name: PropTypes.string,
+  tableFieldIndex: PropTypes.number,
+  type: PropTypes.oneOf(Object.values(FEATURE_TYPE)),
+  domain: PropTypes.arrayOf(
+    PropTypes.oneOfType([PropTypes.string, PropTypes.number])
+  ),
+});
+
+export const FILTER = PropTypes.shape({
+  name: PropTypes.string.isRequired,
+  key: PropTypes.number.isRequired,
+  value: PropTypes.oneOfType([
+    PropTypes.number,
+    PropTypes.string,
+    PropTypes.func,
+    PropTypes.arrayOf(PropTypes.number),
+    PropTypes.arrayOf(PropTypes.string),
+    PropTypes.arrayOf(PropTypes.bool),
+  ]).isRequired,
+  type: PropTypes.oneOf(Object.values(FILTER_TYPE)),
+});
+
+export const STATE_DATA_TYPES = {
+  modelsMeta: PropTypes.shape({
+    nClasses: PropTypes.number,
+    classLabels: PropTypes.arrayOf(
+      PropTypes.oneOfType([PropTypes.string, PropTypes.number])
+    ),
+  }),
+  metric: PropTypes.shape({
+    name: PropTypes.string.isRequired,
+    description: PropTypes.string,
+    func: PropTypes.func.isRequired,
+  }),
+  isManualSegmentation: PropTypes.bool,
+  baseCols: PropTypes.arrayOf(PropTypes.number),
+  nClusters: PropTypes.number,
+  /** input segment filters, each of shape key, value, and type */
+  segmentFilters: PropTypes.arrayOf(PropTypes.arrayOf(FILTER)),
+  segmentGroups: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)),
+};
```

Comment:
Please double-check if any of the above fields need to have `.isRequired`.
Seems we don't provide defaultProps for `FIELD`,  `FILTER` when using them in the components.
And it's highly likely we won't catch the errors if developers failed to follow the expected form.

Reply:
My convention here is that in the bottom level component, such as `segment-filter.js`, marks necessary fields as `isRequired` such that it will report error whenever `SegmentFilter` needs to be rendered but no valid filter data is passed in; in non-bottom level components, such as `segment-group-panel.js`, provide empty defaults such that no error will be caused if a developer tries to perform array operations when no array is passed in (but `segment-group-panel.js` will not check filter object types)